src/app/Trade.ts

```diff
@@ -1,84 +1,163 @@
+    this.disposeResult(offer, playerHerd, target);
+    return true;
   }
 
   private areOffersInvalid(
-    [animal, count]: Offer,
+    offer: Offer[],
     playerHerd: Herd,
-    [bankAnimal, bankCount]: Offer,
+    target: Offer[],
+  ): boolean {
+    return (
+      (offer.length > 1 && target.length > 1) ||
+      this.isOfferQuantityInvalid(offer, playerHerd) ||
+      this.isOfferQuantityInvalid(target, this.bank.theHerd) ||
+      this.areQuantityOnSidesInvalid(offer, target)
+    );
+  }
+
+  private isOfferQuantityInvalid(
+    offer: Offer[],
+    herd: Herd,
   ): boolean {
     return (
-      (count > 1 && bankCount > 1) ||
-      playerHerd.getAnimalNumber(animal) < count ||
-      this.bank.theHerd.getAnimalNumber(bankAnimal) < bankCount
+      offer.filter(
+        ([animal, count]) => herd.getAnimalNumber(animal) >= count,
+      ).length !== offer.length
     );
   }
 
-  private calculateValue([animal, count]: Offer): number {
-    return multiply(
-      ConvertAnimalName.toAnimalObject(animal).theValue,
-      count,
+  private areQuantityOnSidesInvalid(
+    offer: Offer[],
+    target: Offer[],
+  ): boolean {
+    return !(
+      (offer.length === 1 && offer[0][1] === 1) ||
+      (target.length === 1 && target[0][1] === 1)
     );
   }
 
-  private adjustOffer(offer: Offer, target: Offer): void {
-    offer[1] -= 1;
-    if (this.calculateValue(offer) <= this.calculateValue(target)) {
+  private getValue(animalName: AnimalNames): number {
+    return [...this.livestockConfig, ...this.protectorsConfig].filter(
+      (animal) => animal.name === animalName,
+    )[0].tradeValue;
+  }
+
+  private calculateValue(offer: Offer[]): number {
+    return sum(
+      offer.map(([animal, count]) =>
+        multiply(this.getValue(animal), count),
+      ),
+    );
+  }
+
+  private containsOneItem(offer: Offer[]): boolean {
+    return offer.length === 1 && offer[0][1] === 1;
+  }
+
+  private calculateBankValue(): number {
+    return sum(
+      this.bank.theHerd.theAnimals.map(([animal, count]) =>
+        multiply(animal.theValue, count),
+      ),
+    );
+  }
+
+  /**
+   * Adjusts player offer for exchange in bank
+   * @param offer accepts array of tuple
+   * @param value accepts number equal to value of offer
+   * @param price accepts number equal to desired target
+   */
+  private adjustOffer(
+    offer: Offer[],
+    value: number,
+    price: number,
+  ): void {
+    const valueOfFirstItem = this.calculateValue([offer[0]]);
+    const differece = value - price;
+    if (valueOfFirstItem === differece) {
+      offer.shift();
+      return;
+    }
+    if (valueOfFirstItem < differece) {
+      offer.shift();
+      this.adjustOffer(offer, this.calculateValue(offer), price);
       return;
     }
-    this.adjustOffer(offer, target);
+    const singleAnimalCost = valueOfFirstItem / offer[0][1];
+    const animalsToRemoveFromOffer = differece / singleAnimalCost;
```

Comment:
If you want, you could correct the difference spelling (differece)

Reply:
Will be corrected after merge with other task.